# PFGPred: A stack ensemble classifier for the detection of fusion genes in plants from RNA-seq data
PFGPred (Plant Fusion Gene Predictor) is a comprehensive bioinformatics pipeline for the identification of fusion genes in plants using RNA-Seq data. 
It integrates conventional fusion detection tools, a feature extraction framework, and a stacked ensemble machine-learning model specifically trained on plant datasets to identify high-confidence fusion events with significantly reduced false positives.

# PFGPred operates in two major steps:
**1. Fusion Detection & Feature Extraction:** RNA-Seq data are processed using conventional fusion callers to detect candidate fusion transcripts. For each candidate, a set of features derived from RNA-Seq and genomic annotation was extracted.

**2.Classification of True vs. False Fusions:** Extracted features were evaluated using a plant-optimized stacked ensemble model (XGBoost, Random Forest, and LSTM with a logistic regression meta-classifier) to distinguish true fusion transcripts from false positives. This ensures the identification of biologically relevant, high-confidence fusion genes.

PFGPred thus provides an end-to-end, accurate, and plant-specific framework for fusion gene discovery without the need for validation on WGS data.

**Key features of PFGPred:**

-**Integrating RNA-Seq fusion predictions with machine learning**

-**Trained on fusion events validated on WGS datasets**

-**Supports any RNA-Seq fusion caller**

-**Allows species-specific retraining**

# Installation
1. **Clone the repository**:
```bash
git clone https://github.com/skbinfo/PFGPred.git
```
2. **Create Environment**:
```bash
cd PFGPred
chmod +x setup.sh
./setup.sh
```
3. **Activate Environment**:
```bash
conda activate PFGPred_env
```

# Running the Pipeline
1. **Prepare the Configuration File**

Edit config.yaml and provide paths to the required reference files:
```bash
Reference genome (.fasta)
Genome annotation (.gtf)
CTAT resource library (required for STAR-Fusion)
```

2. **How to Prepare a CTAT Genome Library**

PFGPred uses STAR-Fusion as its default fusion detection tool, which requires a CTAT Genome Resource Library built for your species.
You can prepare this library by following the official CTAT tutorial:

**ðŸ”— CTAT Genome Lib Preparation Guide**
 https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/wiki


3. **Specify Samples**

List the sample identifiers in one of the following files based on your sample type:
```
single.txt for single-end reads
paired.txt for paired-end reads
```
Each line should contain one sample ID (matching FASTQ file names).


4. **Create Output Directory**
```bash
mkdir Fusion_output
```


5. **Organize Input Files**

Place all input FASTQ files into the directory where ft.py is located.
Run Feature Extraction
```bash
python ft.py \
--start step1 \
--config config.yaml \
â€“gtf path/to/annotation.gtf
```

This step:
- ***Processes outputs from fusion callers (Default: STAR-Fusion)***
-***Extracts RNA-Seq based features for each fusion transcript***
-***Generates a feature table compatible with the model***


6. **Run the Prediction Model**
To classify fusions using the pretrained stacked ensemble model:
```bash

```
***Or upload the generated feature table to the PFGPred Web Server (http://223.31.159.15/PFGPred/predict.php) if the number of entries in the file is less than 10000.***


The output will provide:


1.**probability scores**


2.**classification labels**




3.**high-confidence filtered fusion list (Default threshold 0.9)**


# Kickstart Mode (Using Outputs from Any Fusion Caller)


PFGPred also supports a Kickstart Mode, allowing you to use fusion predictions generated by any RNA-Seq fusion detection tool. This mode is useful when STAR-Fusion is not preferred or when you already have fusion caller outputs from another workflow. In Kickstart mode, you simply provide the output of your fusion caller to the feature extraction module and then run the PFGPred model.

1. **Input Format**

Prepare the fusion caller output in the following format. If any field is missing or not available, use **. (dot)** or **NF** to indicate missing information. This ensures consistent parsing during feature extraction.

| Sample_ID | #FusionName | 5_geneid | 3_geneid | FFPM | CDS_LEFT_ID | CDS_RIGHT_ID | Splice_Pattern | Chromosome1 | LeftBreakpoint | LeftStrand | Chromosome2 | RightBreakpoint | RightStrand | Splice_Site | Total_Count_(SC+RC) | Splice_Pattern_Class |
| :-------- | :---------- | :------- | :------- | :---- | :----------- | :------------ | :-------------- | :----------- | :-------------- | :---------- | :----------- | :--------------- | :----------- | :----------- | :-------------------- | :--------------------- |
| ERR1588794 | AT2G47640->AT4G18230 | AT2G47640 | AT4G18230 | 2.1745 | . | . | . | 2 | 19537529 | #ERROR! | 4 | 10081731 | - | GT-AG | 18 | Canonical |
| ERR1588794 | AT2G37600->AT5G02450 | AT2G37600 | AT5G02450 | 2.1746 | transcript:AT2G37600.1 | transcript:AT5G02450.1 | INFRAME | 2 | 15774593 | - | 5 | 534358 | #ERROR! | GT-AG | 18 | Canonical |
| ERR1588794 | AT3G23050->AT2G22670 | AT3G23050 | AT2G22670 | 0.2416 | transcript:AT3G23050.1 | transcript:AT2G22670.1 | INFRAME | 3 | 8196210 | #ERROR! | 2 | 9638275 | #ERROR! | GT-AG | 2 | Canonical |
| SRR3458666 | AT3G53870->AT5G35530 | AT3G53870 | AT5G35530 | 1.5013 | transcript:AT3G53870.1 | transcript:AT5G35530.1 | INFRAME | 3 | 19952063 | #ERROR! | 5 | 13710916 | - | AT-AC | 6 | Canonical |
| SRR3458666 | AT4G27040->AT4G27080 | AT4G27040 | AT4G27080 | 0.7507 | transcript:AT4G27040.1 | transcript:AT4G27080.1 | INFRAME | 4 | 13573491 | - | 4 | 13591228 | #ERROR! | GT-AG | 3 | Canonical |
| SRR3458666 | AT2G24060->AT1G56350 | AT2G24060 | AT1G56350 | 0.7507 | transcript:AT2G24060.1 | transcript:AT1G56350.1 | INFRAME | 2 | 10230006 | #ERROR! | 1 | 21097040 | #ERROR! | AT-AG | 3 | Non-Canonical |
| SRR3458666 | AT1G60800->AT1G04880 | AT1G60800 | AT1G04880 | 0.7507 | . | . | . | 1 | 22385933 | - | 1 | 1378266 | - | GT-AG | 3 | Canonical |
| SRR3458666 | AT1G71830->AT2G13790 | AT1G71830 | AT2G13790 | 0.5004 | transcript:AT1G71830.1 | transcript:AT2G13790.1 | FRAMESHIFT | 1 | 27019380 | #ERROR! | 2 | 5743213 | #ERROR! | GT-AG | 2 | Canonical |

